agent/commands/run.go

```diff
@@ -56,75 +55,85 @@ func Run() {
 		s := <-signals
 		signal.Stop(signals)
 		l.Warnf("Got %s, shutting down...", unix.SignalName(s.(unix.Signal))) //nolint:forcetypeassert
-		if cfg != nil {
-			cleanupTmp(cfg.Paths.TempDir, l)
-		}
-		cancel()
+		rootCancel()
 	}()
 
-	configStorage := config.NewStorage(nil)
-	configFilepath, err := configStorage.Reload(l)
-	if err != nil {
-		l.Fatalf("Failed to load configuration: %s.", err)
-	}
-
-	cfg = configStorage.Get()
-
-	cleanupTmp(cfg.Paths.TempDir, l)
-	connectionUptimeService := connectionuptime.NewService(cfg.WindowConnectedTime)
-	connectionUptimeService.RunCleanupGoroutine(ctx)
-	v := versioner.New(&versioner.RealExecFunctions{})
-	supervisor := supervisor.NewSupervisor(ctx, v, configStorage)
-	connectionChecker := connectionchecker.New(configStorage)
-	r := runner.New(cfg.RunnerCapacity)
-	client := client.New(configStorage, supervisor, r, connectionChecker, v, connectionUptimeService, logStore)
-	localServer := agentlocal.NewServer(configStorage, supervisor, client, configFilepath, logStore)
-
-	var wg sync.WaitGroup
-	wg.Add(3)
-	go func() {
-		defer wg.Done()
-		supervisor.Run(ctx)
-		cancel()
-	}()
-	go func() {
-		defer wg.Done()
-		r.Run(ctx)
-		cancel()
-	}()
-	go func() {
-		defer wg.Done()
-		localServer.Run(ctx)
-		cancel()
-	}()
+	configStorage, configFilepath := prepareConfig(l)
 
 	for {
-		_, err = configStorage.Reload(l)
-		if err != nil {
-			l.Fatalf("Failed to load configuration: %s.", err)
-		}
-
+		ctx, cancel := context.WithCancel(rootCtx)
 		cfg := configStorage.Get()
 
-		config.ConfigureLogger(cfg)
-		logStore.Resize(cfg.LogLinesCount)
-		l.Debugf("Loaded configuration: %+v", cfg)
+		prepareLogger(cfg, logStore, l)
+
+		v := versioner.New(&versioner.RealExecFunctions{})
+		supervisor := supervisor.NewSupervisor(ctx, v, configStorage)
+		connectionChecker := connectionchecker.New(configStorage)
+		r := runner.New(cfg.RunnerCapacity)
+		client := client.New(configStorage, supervisor, r, connectionChecker, v, prepareConnectionService(ctx, cfg), logStore)
+		localServer := agentlocal.NewServer(configStorage, supervisor, client, configFilepath, logStore)
 
 		logrus.Infof("Window check connection time is %.2f hour(s)", cfg.WindowConnectedTime.Hours())
-		connectionUptimeService.SetWindowPeriod(cfg.WindowConnectedTime)
 
-		clientCtx, cancelClientCtx := context.WithCancel(ctx)
+		var wg sync.WaitGroup
+		wg.Add(3)
+		reloadCh := make(chan bool, 1)
+		go func() {
+			defer wg.Done()
+			supervisor.Run(ctx)
+			cancel()
+		}()
+		go func() {
+			defer wg.Done()
+			r.Run(ctx)
+			cancel()
+		}()
+		go func() {
+			defer wg.Done()
+			localServer.Run(ctx, reloadCh)
+			cancel()
+		}()
+
+		action := ""
+		for action == "" {
+			clientCtx, cancelClientCtx := context.WithCancel(ctx)
+			client.Run(clientCtx)
+
+			cancelClientCtx()
+			<-client.Done()
+
+			select {
+			case <-reloadCh:
+				action = "reload"
+			default:
+				if ctx.Err() != nil {
+					l.Debug(ctx.Err())
+					action = "done"
+				}
```

Comment:
I think we can extract it to a separate function and in that case, it will be easier to handle

Reply:
Great idea how to simplify it. Extracted into method. 

About just ctx.Done() here its problem for reconnect, its waiting. In case of just checking error it can continue in case of no error yet.